src/main/java/net/minecraftforge/common/tags/ForgeTagManager.java

```diff
@@ -0,0 +1,299 @@
+        registerForgeTags(GameData.ENCHANTMENTS, ENCHANTMENTS, (manager -> Tags.Enchantments.setEnchantmentTagCollection((ForgeTagCollection<Enchantment>) manager.getTagsForId(GameData.ENCHANTMENTS))), "enchantment");
+        registerForgeTags(GameData.PROFESSIONS, VILLAGER_PROFESSIONS, (manager -> Tags.VillagerProfessions.setVillagerProfessionTagCollection((ForgeTagCollection<VillagerRegistry.VillagerProfession>) manager.getTagsForId(GameData.PROFESSIONS))), "villager profession");
+        registerForgeTags(GameData.ENTITIES, ENTITIES, (manager -> Tags.Entities.setEntityTagCollection((ForgeTagCollection<Entity>) manager.getTagsForId(GameData.ENTITIES))), "entity");
+        registerForgeTags(GameData.TILEENTITIES, TILE_ENTITIES, (manager -> Tags.TileEntities.setTileEntityTagCollection((ForgeTagCollection<TileEntity>) manager.getTagsForId(GameData.TILEENTITIES))), "tile entity");
+    }
+
+    private static int generation = 0;
+
+    public static int getGeneration()
+    {
+        return generation;
+    }
+
+    private Map<ResourceLocation, TagCollectionValueEntry> tagCollections;
+
+    public ForgeTagManager()
+    {
+        super();
+        tagCollections = Maps.newHashMapWithExpectedSize(factories.size());
+        for (Map.Entry<ResourceLocation, TagCollectionFactoryEntry> entry : factories.entrySet())
+            tagCollections.put(entry.getKey(), entry.getValue().create(entry.getKey(), this));
+        tagCollections = Collections.unmodifiableMap(tagCollections);
+    }
+
+
+    public ForgeTagCollection<?> getTagsForId(ResourceLocation id)
+    {
+        return tagCollections.get(id).getTags();
+    }
+
+    @Override
+    public ForgeTagCollection<Block> getBlocks()
+    {
+        return (ForgeTagCollection<Block>) super.getBlocks();
+    }
+
+    @Override
+    public ForgeTagCollection<Item> getItems()
+    {
+        return (ForgeTagCollection<Item>) super.getItems();
+    }
+
+    @Override
+    public ForgeTagCollection<Fluid> getFluids()
+    {
+        return (ForgeTagCollection<Fluid>) super.getFluids();
+    }
+
+    @Override
+    public void clear()
+    {
+        for (Map.Entry<ResourceLocation, TagCollectionValueEntry> entry : tagCollections.entrySet())
+            entry.getValue().clear();
+    }
+
+    @Override
+    public void onResourceManagerReload(IResourceManager resourceManager)
+    {
+        for (Map.Entry<ResourceLocation, TagCollectionValueEntry> entry : tagCollections.entrySet())
+        {
+            entry.getValue().clear();
+            entry.getValue().reload(entry.getKey(), resourceManager, this);
+        }
+        generation++;
+    }
+
+    @Override
+    public void write(PacketBuffer buffer)
+    {
+        this.getBlocks().write(buffer);
+        this.getItems().write(buffer);
+        this.getFluids().write(buffer);
```

Comment:
Welcome to the world of limited packet sizes. This will explode and explode fast. Do not hijack the vanilla packet.

Reply:
Ok, so should I hook into the Forge Indexed Messages (I think that's how the Handschake Messages used to transferring Registry Data in 1.12 were called)?